File: 2022/05/15/binaryTreePaths.py
"""
https://leetcode.com/problems/invert-binary-tree/
Author: Andrew Jarombek
Date: 5/15/2022
"""

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        self.paths = []
        self.dfs(root, "", "")
        return self.paths

    def dfs(self, root: Optional[TreeNode], path: str, sep: str) -> None:
        path = f"{path}{sep}{root.val}"
        sep = "->"

        if root.left is None and root.right is None:
            self.paths.append(path)

        if root.left is not None:
            self.dfs(root.left, path, sep)

        if root.right is not None:
            self.dfs(root.right, path, sep)

File: 2023/02/03/go/binary_tree_paths.go
/**
 * https://leetcode.com/problems/binary-tree-paths/
 * Author: Andrew Jarombek
 * Date: 2/3/2023
 */

package _go

import "strconv"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
	var result []string
	dfs(root, "", &result)
	return result
}

func dfs(node *TreeNode, path string, result *[]string) {
	if node.Left == nil && node.Right == nil {
		if len(path) > 0 {
			path += "->"
		}
		*result = append(*result, path+strconv.Itoa(node.Val))
		return
	}

	if len(path) > 0 {
		path = path + "->"
	}

	if node.Left != nil {
		dfs(node.Left, path+strconv.Itoa(node.Val), result)
	}

	if node.Right != nil {
		dfs(node.Right, path+strconv.Itoa(node.Val), result)
	}
}
